File: lib/strava.js
// lib/strava.js
// Handles token refresh and authenticated requests to the Strava API.
// Strava access tokens expire every 6h — this refreshes on every cold start
// using the long-lived refresh token stored in env vars.

let cachedToken = null;
let cachedExpiry = 0;

async function getAccessToken() {
  const now = Math.floor(Date.now() / 1000);
  if (cachedToken && now < cachedExpiry - 60) {
    return cachedToken;
  }

  const res = await fetch("https://www.strava.com/oauth/token", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      client_id: process.env.STRAVA_CLIENT_ID,
      client_secret: process.env.STRAVA_CLIENT_SECRET,
      grant_type: "refresh_token",
      refresh_token: process.env.STRAVA_REFRESH_TOKEN
    })
  });

  if (!res.ok) {
    throw new Error(`Falha ao renovar token Strava: ${res.status} ${await res.text()}`);
  }

  const data = await res.json();
  cachedToken = data.access_token;
  cachedExpiry = data.expires_at;

  // NOTE: Strava may rotate the refresh_token. In production, persist
  // data.refresh_token somewhere durable (Vercel KV, a small DB, etc.)
  // instead of relying solely on the static env var. See README.

  return cachedToken;
}

async function stravaGet(path) {
  const token = await getAccessToken();
  const res = await fetch(`https://www.strava.com/api/v3${path}`, {
    headers: { Authorization: `Bearer ${token}` }
  });
  if (!res.ok) {
    throw new Error(`Strava API ${path} -> ${res.status} ${await res.text()}`);
  }
  return res.json();
}

module.exports = { getAccessToken, stravaGet };

File: lib_py/strava.py
# lib_py/strava.py
# Python port of lib/strava.js — handles Strava OAuth token refresh and
# authenticated requests. Uses the same STRAVA_CLIENT_ID / STRAVA_CLIENT_SECRET
# / STRAVA_REFRESH_TOKEN env vars already configured for the Node backend.

import os
import time
import requests

_cached_token = None
_cached_expiry = 0


def get_access_token():
    global _cached_token, _cached_expiry
    now = int(time.time())
    if _cached_token and now < _cached_expiry - 60:
        return _cached_token

    res = requests.post(
        "https://www.strava.com/oauth/token",
        json={
            "client_id": os.environ.get("STRAVA_CLIENT_ID"),
            "client_secret": os.environ.get("STRAVA_CLIENT_SECRET"),
            "grant_type": "refresh_token",
            "refresh_token": os.environ.get("STRAVA_REFRESH_TOKEN"),
        },
    )
    if not res.ok:
        raise RuntimeError(f"Failed to refresh Strava token: {res.status_code} {res.text}")

    data = res.json()
    _cached_token = data["access_token"]
    _cached_expiry = data["expires_at"]
    # NOTE: same caveat as the Node client — Strava may rotate the refresh
    # token; persist data["refresh_token"] somewhere durable for production use.
    return _cached_token


def strava_get(path, params=None):
    token = get_access_token()
    res = requests.get(
        f"https://www.strava.com/api/v3{path}",
        headers={"Authorization": f"Bearer {token}"},
        params=params or {},
    )
    if not res.ok:
        raise RuntimeError(f"Strava API {path} -> {res.status_code} {res.text}")
    return res.json()
